drivers/sensor/bosch/bmp180/bmp180.c

```diff
@@ -0,0 +1,463 @@
+
+#include "bmp180.h"
+
+#define DT_DRV_COMPAT bosch_bmp180
+
+LOG_MODULE_REGISTER(BMP180, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Calibration Registers structure */
+struct bmp180_cal_data {
+	int16_t ac1;
+	int16_t ac2;
+	int16_t ac3;
+	uint16_t ac4;
+	uint16_t ac5;
+	uint16_t ac6;
+	int16_t b1;
+	int16_t b2;
+	int16_t mb;
+	int16_t mc;
+	int16_t md;
+} __packed;
+
+struct bmp180_config {
+	const struct i2c_dt_spec i2c;
+};
+
+struct bmp180_data {
+	uint8_t osr_pressure;
+	int32_t raw_press;
+	int32_t raw_temp;
+	int32_t comp_temp;
+	struct bmp180_cal_data cal;
+};
+
+static inline int bmp180_bus_check(const struct device *dev)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return i2c_is_ready_dt(&cfg->i2c) ? 0 : -ENODEV;
+}
+
+static inline int bmp180_reg_read(const struct device *dev, uint8_t start,
+				 uint8_t *buf, int size)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return i2c_burst_read_dt(&cfg->i2c, start, buf, size);
+}
+
+static inline int bmp180_reg_write(const struct device *dev, uint8_t reg,
+				   uint8_t val)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return i2c_reg_write_byte_dt(&cfg->i2c, reg, val);
+}
+
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+static int bmp180_attr_set_oversampling(const struct device *dev,
+					enum sensor_channel chan, uint16_t val)
+{
+	struct bmp180_data *data = dev->data;
+
+	/* Value must be a positive value 0-3 */
+	if ((chan != SENSOR_CHAN_PRESS) || (val > 3)) {
+		return -EINVAL;
+	}
+
+	data->osr_pressure = val;
+
+	return 0;
+}
+#endif /* CONFIG_BMP180_OSR_RUNTIME */
+
+static int bmp180_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	int ret;
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif /* CONFIG_PM_DEVICE */
+
+	switch (attr) {
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+	case SENSOR_ATTR_OVERSAMPLING:
+		ret = bmp180_attr_set_oversampling(dev, chan, val->val1);
+		break;
+#endif /* CONFIG_BMP180_OSR_RUNTIME */
+
+	default:
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+
+static int read_raw_temperature(const struct device *dev)
+{
+	int ret;
+	uint8_t ctrlreg;
+	uint8_t reg16[2];
+	struct bmp180_data *data = dev->data;
+
+	/* send temperature measurement command */
+	ret = bmp180_reg_write(dev, BMP180_REG_MEAS_CTRL, BMP180_CMD_GET_TEMPERATURE);
+	if (ret) {
+		return ret;
+	}
+
+	/*
+	 * for the first while read get_temp_delay+1ms which is the convension time
+	 * descripted in the data-sheet in case the register not yet ready wait again
+	 */
+	k_sleep(K_MSEC(BMP180_CMD_GET_TEMP_DELAY));
+
+	while (true) {
+		k_sleep(K_MSEC(1));
+		ret = bmp180_reg_read(dev, BMP180_REG_MEAS_CTRL, &ctrlreg, 1);
+		if (ret) {
+			return ret;
+		}
+
+		/* bit 5 is 1 until data registers are ready */
+		if ((ctrlreg & BMP180_STATUS_CMD_RDY) == 0) {
+			break;
+		}
+	}
```

Comment:
This feels a bit dangerous to just keep looping, can we have a Kconfig for max retries? Maybe default to something like 5?

Reply:
The datasheet said, "maximum conversion time" and bit5 will set when ready. For me this is reliable, not?
In my testing never had once a retry, checked with a printk().
The cheapest I would do is uint8_t retry=2;  --retry;  if (retry==0) { errormsg & -ETIMEOUT }; and not bloat the Kconfig

Reply:
I'm OK with a constants retry value instead of a Kconfig.